commit c5590a6821e37f3b29735f55eb0c2b9c0924138c
Author: Steffan Karger <steffan.karger@fox-it.com>
Date:   Thu Nov 20 13:43:05 2014 +0100

    Drop too-short control channel packets instead of asserting out.
    
    This fixes a denial-of-service vulnerability where an authenticated client
    could stop the server by triggering a server-side ASSERT().
    
    OpenVPN would previously ASSERT() that control channel packets have a
    payload of at least 4 bytes. An authenticated client could trigger this
    assert by sending a too-short control channel packet to the server.
    
    Thanks to Dragana Damjanovic for reporting the issue.
    
    This bug has been assigned CVE-2014-8104.
    
    Signed-off-by: Steffan Karger <steffan.karger@fox-it.com>
    Acked-by: Gert Doering <gert@greenie.muc.de>
    Message-Id: <1CED409804E2164C8104F9E623B08B9018803B0FE7@FOXDFT02.FOX.local>
    Signed-off-by: Gert Doering <gert@greenie.muc.de>

diff --git a/src/openvpn/ssl.c b/src/openvpn/ssl.c
index 2adfa26..cdc8eb1 100644
--- ssl.c
+++ ssl.c
@@ -3751,7 +3751,11 @@
   ALLOC_ARRAY_CLEAR_GC (options, char, TLS_OPTIONS_LEN, &gc);
 		  
   /* discard leading uint32 */
-  ASSERT (buf_advance (buf, 4));
+  if (!buf_advance (buf, 4)) {
+    msg (D_TLS_ERRORS, "TLS ERROR: Plaintext buffer too short (%d bytes).",
+	buf->len);
+    goto error;
+  }
 
   /* get key method */
   key_method_flags = buf_read_u8 (buf);
